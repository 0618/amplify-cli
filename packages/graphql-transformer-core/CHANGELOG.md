# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.0.9"></a>
## [1.0.9](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.8...graphql-transformer-core@1.0.9) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.8"></a>
## [1.0.8](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.7...graphql-transformer-core@1.0.8) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.7"></a>
## [1.0.7](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.6...graphql-transformer-core@1.0.7) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.6"></a>
## [1.0.6](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.5...graphql-transformer-core@1.0.6) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.5"></a>
## [1.0.5](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.4...graphql-transformer-core@1.0.5) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.4"></a>
## 1.0.4 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.3"></a>
## 1.0.3 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.2"></a>
## 1.0.2 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.1"></a>
## 1.0.1 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core
