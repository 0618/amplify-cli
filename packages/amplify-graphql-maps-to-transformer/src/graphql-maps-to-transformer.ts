import { TransformerPluginBase, InvalidDirectiveError } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerPluginType,
  TransformerSchemaVisitStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { ObjectTypeDefinitionNode, DirectiveNode } from 'graphql';
import { attachInputMappingSlot, attachResponseMappingSlot } from './field-mapping-resolvers';

const directiveName = 'mapsTo';

const directiveDefinition = `
  directive @${directiveName}(name: String!) on OBJECT
`;

export class MapsToTransformer extends TransformerPluginBase {
  constructor() {
    super(`amplify-maps-to-transformer`, directiveDefinition, TransformerPluginType.GENERIC);
  }

  object = (definition: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerSchemaVisitStepContextProvider) => {
    const modelName = definition.name.value;
    const prevNameNode = directive.arguments?.find(arg => arg.name.value === 'name');
    if (!prevNameNode) {
      throw new InvalidDirectiveError(`name is required in @${directiveName} directive`);
    }

    if (prevNameNode.value.kind !== 'StringValue') {
      throw new InvalidDirectiveError(`A single string must be provided for "name" in @${directiveName} directive`);
    }

    const originalName = prevNameNode.value.value;
    ctx.resourceHelper.setModelNameMapping(modelName, originalName);
  };

  generateResolvers = (context: TransformerContextProvider) =>
    context.resourceHelper.getResolverMapRegistry().forEach(({ resolverTypeName, resolverFieldName, fieldMap, isResultList }) => {
      const resolver = context.resolvers.getResolver(resolverTypeName, resolverFieldName);
      if (!resolver) {
        return;
      }
      if (resolverTypeName === 'Mutation') {
        attachInputMappingSlot({ resolver, resolverFieldName, resolverTypeName, fieldMap });
        attachResponseMappingSlot({ slotName: 'postUpdate', resolver, resolverFieldName, resolverTypeName, fieldMap, isList: false });
        return;
      }
      attachResponseMappingSlot({
        slotName: 'postDataLoad',
        resolver,
        resolverFieldName,
        resolverTypeName,
        fieldMap: fieldMap,
        isList: isResultList,
      });
    });
}
