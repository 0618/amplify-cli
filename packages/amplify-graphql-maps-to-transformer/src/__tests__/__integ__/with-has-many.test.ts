import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { GraphQLTransform } from '@aws-amplify/graphql-transformer-core';
import { BelongsToTransformer, HasManyTransformer } from '@aws-amplify/graphql-relational-transformer';
import { MapsToTransformer } from '../../graphql-maps-to-transformer';

const mappedHasMany = /* GraphQL */ `
  type Employee @model @mapsTo(name: "Person") {
    id: ID!
    tasks: [Task] @hasMany
  }

  type Task @model {
    id: ID!
    title: String
  }
`;

const transformSchema = (schema: string) => {
  const transformer = new GraphQLTransform({
    transformers: [new ModelTransformer(), new HasManyTransformer(), new BelongsToTransformer(), new MapsToTransformer()],
    sandboxModeEnabled: true,
  });
  return transformer.transform(schema);
};

describe('@mapsTo with @hasMany', () => {
  it('adds CRUD input and output mappings on related type and maps related type in hasMany field resolver', () => {
    const out = transformSchema(mappedHasMany);
    const expectedResolvers: string[] = [
      'Mutation.createTask.postUpdate.1.res.vtl',
      'Mutation.createTask.init.2.req.vtl',
      'Mutation.updateTask.postUpdate.1.res.vtl',
      'Mutation.updateTask.init.2.req.vtl',
      'Query.getTask.postDataLoad.1.res.vtl',
      'Query.listTasks.postDataLoad.1.res.vtl',
      'Employee.tasks.postDataLoad.1.res.vtl',
    ];
    expectedResolvers.forEach(resolver => {
      expect(out.resolvers[resolver]).toMatchSnapshot();
    });
  });
});
