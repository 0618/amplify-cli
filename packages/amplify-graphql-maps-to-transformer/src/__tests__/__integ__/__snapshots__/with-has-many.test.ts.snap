// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 1`] = `
"$util.qr($ctx.prev.result.put(\\"employeeTasksId\\", $ctx.prev.result.personTasksId))
$util.qr($ctx.prev.result.remove(\\"personTasksId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 2`] = `
"$util.qr($ctx.args.input.put(\\"personTasksId\\", $ctx.args.input.employeeTasksId))
$util.qr($ctx.args.input.remove(\\"employeeTasksId\\"))
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 3`] = `
"$util.qr($ctx.prev.result.put(\\"employeeTasksId\\", $ctx.prev.result.personTasksId))
$util.qr($ctx.prev.result.remove(\\"personTasksId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 4`] = `
"$util.qr($ctx.args.input.put(\\"personTasksId\\", $ctx.args.input.employeeTasksId))
$util.qr($ctx.args.input.remove(\\"employeeTasksId\\"))
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 5`] = `
"$util.qr($ctx.prev.result.put(\\"employeeTasksId\\", $ctx.prev.result.personTasksId))
$util.qr($ctx.prev.result.remove(\\"personTasksId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 6`] = `
"#foreach( $item in $ctx.prev.result.items )
  $util.qr($item.put(\\"employeeTasksId\\", $item.personTasksId))
  $util.qr($item.remove(\\"personTasksId\\"))
#end
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 7`] = `
"#foreach( $item in $ctx.prev.result.items )
  $util.qr($item.put(\\"employeeTasksId\\", $item.personTasksId))
  $util.qr($item.remove(\\"personTasksId\\"))
#end
$util.toJson($ctx.prev.result)"
`;
